File: interleaving-string.java
public class Solution {
    public boolean isInterleave(String s1, String s2, String s3) {
        if (s1.length() + s2.length() != s3.length()) return false;
        boolean[][] interleave = new boolean[s1.length() + 1][s2.length() + 1];
        interleave[0][0] = true;
        for (int i = 0; i <= s1.length(); i++) {
            for (int j = 0; j <= s2.length(); j++) {
                if (!interleave[i][j]) continue;
                if (i < s1.length() && s1.charAt(i) == s3.charAt(i + j))
                    interleave[i + 1][j] = true;
                if (j < s2.length() && s2.charAt(j) == s3.charAt(i + j))
                    interleave[i][j + 1] = true;
            }
        }
        return interleave[s1.length()][s2.length()];
    }
}

File: interleaving-string.cc
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
  bool isInterleave(string s1, string s2, string s3) {
    if (s1.size() + s2.size() != s3.size()) return false;
    vector<vector<bool> > interleaving(s1.size() + 1,
        vector<bool>(s2.size() + 1));
    interleaving[0][0] = true;
    for (int i = 0; i <= s1.size(); i++) {
      for (int j = 0; j <= s2.size(); j++) {
        if (!interleaving[i][j]) continue;
        if (i < s1.size() && s1[i] == s3[i + j]) interleaving[i + 1][j] = true;
        if (j < s2.size() && s2[j] == s3[i + j]) interleaving[i][j + 1] = true;
      }
    }
    return interleaving[s1.size()][s2.size()];
  }
};
